### PR TITLE
 Issue #5: Dynamic Execution Topologies

### PR DESCRIPTION
**Pull request by knute-sifive**
In this PR, Execute is now it's own node and dynamically creates execution pipes based on extensions listed in the
config files:
```
top.cpu.core0.extension.core_extensions:
  execution_topology:
  [["alu", "6"],
   ["fpu", "2"],
   ["br",  "2"]]
```

Work done;
- Added new class `Dispatcher` that is dynamically created by `Dispatch` for each execution pipe defined in the `execution_topology` extension
- Added new class `Execute` that is a node created for the sole purpose of dynamically creating the `ExecutePipe` objects based on the `execution_topology` extension
- Cleaned up old code
- Reduced the `CPUTopology` structure to just static components like Fetch, Decode, Rename, Dispatch, Execute, LS/etc, and Retire
- Updated the arches to include a small, medium, and big core.  Dhrystone trace improves with each architecture, but is limited by LS.

### CODE REVIEW

**Comment by knute-sifive** [comment]
closes #5

File: core/Execute.cpp [hunk 1]
```diff
@@ -1,163 +1,69 @@
...
+            const std::string disp_credits_in = "dispatch.ports.in_" + exe_pipe_rtn->getName() + "_credits";
+            bind_ports(exe_credits_out, disp_credits_in);
+
+            const std::string exe_inst_in   = "execute." + exe_pipe_rtn->getName() + ".ports.in_execute_write";
+            const std::string disp_inst_out = "dispatch.ports.out_" + exe_pipe_rtn->getName() + "_write";
+            bind_ports(exe_inst_in, disp_inst_out);
+
+            const std::string exe_flush_in = "execute." +  exe_pipe_rtn->getName() + ".ports.in_reorder_flush";;
+            const std::string flush_manager = "flushmanager.ports.out_retire_flush";
+            bind_ports(exe_flush_in, flush_manager);
```
Comment (aka2en) [inline core/Execute.cpp:66]: Hm, this is something I'm always wondering about -- bind these units here or at topology level? One concern I come up with is -- that would make the execution unit test have to create dispatch instance.

File: sim/main.cpp [hunk 2]
```diff
@@ -90,7 +93,7 @@ int main(int argc, char **argv)
 
         // Create the simulator
         sparta::Scheduler scheduler;
-        OlympiaSim sim("core_topology_1",
+        OlympiaSim sim("core_topology_simple",
```
Comment (aka2en) [inline sim/main.cpp:96]: I'm lazy -- I like fewer words ...
```suggestion
        OlympiaSim sim("simple",
```

**Review by aka2en — APPROVED** [review]
LGTM

File: core/Execute.cpp (hunk 1 above)
Comment (knute-sifive) [reply core/Execute.cpp:66]: This is true -- I do "hard-code" dispatch and the flush manager in this factory.  I don't like that either.  Perhaps moving this to the topology would make sense...

Comment (knute-sifive) [reply core/Execute.cpp]: Done -- moved to CPUTopology.

**Pull request merged**